fusillade/clouddirectory.py

```diff
@@ -1140,40 +1125,64 @@ def create_policy(self, statement: str) -> str:
         """
         Policy statements follow AWS IAM Policy Grammer. See for grammar details
         https://docs.aws.amazon.com/IAM/latest/UserGuide/reference_policies_grammar.html
         """
-        if not self._statement and self.policy:
-            self._statement = self.cd.get_object_attributes(self.policy,
-                                                            'POLICY',
-                                                            ['policy_document'],
-                                                            self.cd.node_schema
-                                                            )['Attributes'][0]['Value'].popitem()[1].decode("utf-8")
-
-        return self._statement
-
-    @statement.setter
-    def statement(self, statement: str):
-        self._verify_statement(statement)
-        self._set_statement(statement)
+        if policy_type in self.allowed_policy_types:  # check if this policy type is allowed
+            if not self.attached_policies.get(policy_type):  # check if we already have the policy
+                policy_ref = self.cd.get_obj_type_path('policy') + self.get_policy_name(policy_type)
+                try:
+                    resp = self.cd.get_object_attributes(
+                        policy_ref,
+                        'POLICY',
+                        ['policy_document', 'policy_type'],
+                        self.cd.node_schema
+                    )
+                    attrs = dict([(attr['Key']['Name'], attr['Value'].popitem()[1]) for attr in resp['Attributes']])
+                    if attrs['policy_type'] != policy_type:
+                        logger.warning({'message': "Retrieved policy_type does not match requested policy_type.",
+                                        'expected': policy_type,
+                                        'received': attrs['policy_type']
+                                        })
+                    self.attached_policies[policy_type] = attrs['policy_document'].decode("utf-8")
+                except cd_client.exceptions.ResourceNotFoundException:
+                    pass
+            return self.attached_policies.get(policy_type, '')
+        else:
+            FusilladeHTTPException(
+                title='Bad Request',
+                detail=f"{self.object_type} cannot have policy type {policy_type}."
+                f" Allowed types are: {self.allowed_policy_types}")
 
-    def _set_statement(self, statement: str):
-        try:
-            if not self.policy:
-                self.create_policy(statement)
+    def set_policy(self, statement: str, policy_type: str = 'IAMPolicy'):
+        if policy_type in self.allowed_policy_types:
+            try:
+                # check if this object exists
+                self.cd.get_object_information(self.object_ref)
+            except cd_client.exceptions.ResourceNotFoundException:
+                raise FusilladeHTTPException(status=404, title="Not Found", detail="Resource does not exist.")
```

Comment:
Nit: Consider using `requests.codes.not_found` instead of `404`.

Reply:
will fix in the future